File: mobile/src/main/java/com/bayapps/android/robophish/ui/BaseActivity.java
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.bayapps.android.robophish.ui;

import android.annotation.SuppressLint;
import android.app.ActivityManager;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.os.Bundle;
import android.os.RemoteException;
import android.preference.PreferenceManager;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import androidx.annotation.NonNull;

import com.bayapps.android.robophish.MusicService;
import com.bayapps.android.robophish.R;
import com.bayapps.android.robophish.RoboPhishApplicationKt;
import com.bayapps.android.robophish.utils.NetworkHelper;
import com.bayapps.android.robophish.utils.ResourceHelper;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;

import javax.inject.Inject;

import timber.log.Timber;

/**
 * Base activity for activities that need to show a playback control fragment when media is playing.
 */
public abstract class BaseActivity extends ActionBarCastActivity implements MediaBrowserProvider {

    private static final int PLAY_SERVICES_RESOLUTION_REQUEST = 9000;
    private MediaBrowserCompat mMediaBrowser;
    private PlaybackControlsFragment mControlsFragment;
    private BroadcastReceiver mRegistrationBroadcastReceiver;
    private boolean isReceiverRegistered;

    @Inject GoogleApiAvailability googleApiAvailability;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Timber.d("Activity onCreate");

        RoboPhishApplicationKt.inject(this);
        checkPlayServices();

        if (Build.VERSION.SDK_INT >= 21) {
            // Since our app icon has the same color as colorPrimary, our entry in the Recent Apps
            // list gets weird. We need to change either the icon or the color
            // of the TaskDescription.
            ActivityManager.TaskDescription taskDesc = new ActivityManager.TaskDescription(
                    getTitle().toString(),
                    BitmapFactory.decodeResource(getResources(), R.drawable.ic_launcher_white),
                    ResourceHelper.getThemeColor(this, R.attr.colorPrimary,
                            android.R.color.darker_gray));
            setTaskDescription(taskDesc);
        }

        // Connect a media browser just to get the media session token. There are other ways
        // this can be done, for example by sharing the session token directly.
        mMediaBrowser = new MediaBrowserCompat(this,
                new ComponentName(this, MusicService.class), mConnectionCallback, null);

        mRegistrationBroadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                //mRegistrationProgressBar.setVisibility(ProgressBar.GONE);
                SharedPreferences sharedPreferences =
                        PreferenceManager.getDefaultSharedPreferences(context);
            }
        };
    }

    /**
     * Check the device to make sure it has the Google Play Services APK. If
     * it doesn't, display a dialog that allows users to download the APK from
     * the Google Play Store or enable it in the device's system settings.
     */
    private boolean checkPlayServices() {
        int resultCode = googleApiAvailability.isGooglePlayServicesAvailable(this);
        if (resultCode == ConnectionResult.SUCCESS) {
            return true;
        }
        return false;  //just don't start RegistrationIntentService if not available
    }

    public MediaControllerCompat getSupportMediaController() {
        return MediaControllerCompat.getMediaController(this);
    }

    @Override
    protected void onStart() {
        super.onStart();
        Timber.d("Activity onStart");

        mControlsFragment = (PlaybackControlsFragment) getSupportFragmentManager().findFragmentById(R.id.fragment_playback_controls);
        if (mControlsFragment == null) {
            throw new IllegalStateException("Mising fragment with id 'controls'. Cannot continue.");
        }

        hidePlaybackControls();

        mMediaBrowser.connect();
    }

    @Override
    protected void onStop() {
        super.onStop();
        Timber.d("Activity onStop");
        MediaControllerCompat mediaController = MediaControllerCompat.getMediaController(this);
        if (mediaController != null) {
            mediaController.unregisterCallback(mMediaControllerCallback);
        }
        mMediaBrowser.disconnect();
    }

    @Override
    public MediaBrowserCompat getMediaBrowser() {
        return mMediaBrowser;
    }

    protected void onMediaControllerConnected() {
        // empty implementation, can be overridden by clients.
    }

    protected void showPlaybackControls() {
        Timber.d("showPlaybackControls");
        if (NetworkHelper.isOnline(this)) {
            getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(
                    R.animator.slide_in_from_bottom, R.animator.slide_out_to_bottom,
                    R.animator.slide_in_from_bottom, R.animator.slide_out_to_bottom)
                .show(mControlsFragment)
                .commit();
        }
    }

    protected void hidePlaybackControls() {
        Timber.d("hidePlaybackControls");
        getSupportFragmentManager().beginTransaction()
            .hide(mControlsFragment)
            .commit();
    }

    /**
     * Check if the MediaSession is active and in a "playback-able" state
     * (not NONE and not STOPPED).
     *
     * @return true if the MediaSession's state requires playback controls to be visible.
     */
    protected boolean shouldShowControls() {
        MediaControllerCompat mediaController = MediaControllerCompat.getMediaController(this);
        if (mediaController == null ||
            mediaController.getMetadata() == null ||
            mediaController.getPlaybackState() == null) {
            return false;
        }
        switch (mediaController.getPlaybackState().getState()) {
            case PlaybackStateCompat.STATE_ERROR:
            case PlaybackStateCompat.STATE_NONE:
            case PlaybackStateCompat.STATE_STOPPED:
                return false;
            default:
                return true;
        }
    }

    private void connectToSession(MediaSessionCompat.Token token) throws RemoteException {
        MediaControllerCompat mediaController = new MediaControllerCompat(this, token);
        MediaControllerCompat.setMediaController(this, mediaController);
        mediaController.registerCallback(mMediaControllerCallback);

        if (shouldShowControls()) {
            showPlaybackControls();
        } else {
            Timber.d("connectionCallback.onConnected: hiding controls because metadata is null");
            hidePlaybackControls();
        }

        if (mControlsFragment != null) {
            mControlsFragment.onConnected();
        }

        onMediaControllerConnected();
    }

    // Callback that ensures that we are showing the controls
    private final MediaControllerCompat.Callback mMediaControllerCallback =
        new MediaControllerCompat.Callback() {
            @Override
            @SuppressLint("BinaryOperationInTimber")
            public void onPlaybackStateChanged(@NonNull PlaybackStateCompat state) {
                if (shouldShowControls()) {
                    showPlaybackControls();
                } else {
                    Timber.d("mediaControllerCallback.onPlaybackStateChanged: " +
                            "hiding controls because state is %s", state.getState());
                    hidePlaybackControls();
                }
            }

            @Override
            @SuppressLint("BinaryOperationInTimber")
            public void onMetadataChanged(MediaMetadataCompat metadata) {
                if (shouldShowControls()) {
                    showPlaybackControls();
                } else {
                    Timber.d("mediaControllerCallback.onMetadataChanged: " +
                        "hiding controls because metadata is null");
                    hidePlaybackControls();
                }
            }
        };

    private final MediaBrowserCompat.ConnectionCallback mConnectionCallback =
        new MediaBrowserCompat.ConnectionCallback() {
            @Override
            public void onConnected() {
                Timber.d("onConnected");
                try {
                    connectToSession(mMediaBrowser.getSessionToken());
                } catch (RemoteException e) {
                    Timber.e(e, "could not connect media controller");
                    hidePlaybackControls();
                }
            }
        };

}

File: mobile/src/main/java/com/bayapps/android/robophish/ui/BaseActivity.kt
package com.bayapps.android.robophish.ui

import android.app.ActivityManager
import android.content.ComponentName
import android.graphics.BitmapFactory
import android.os.Build
import android.os.Bundle
import androidx.core.content.ContextCompat
import androidx.media3.common.Player
import androidx.media3.session.MediaBrowser
import androidx.media3.session.SessionToken
import com.bayapps.android.robophish.MusicService
import com.bayapps.android.robophish.R
import com.bayapps.android.robophish.utils.NetworkHelper
import com.google.common.util.concurrent.ListenableFuture
import com.google.common.util.concurrent.MoreExecutors
import timber.log.Timber

/**
 * Base activity for activities that need to show a playback control fragment.
 */
abstract class BaseActivity : ActionBarCastActivity(), MediaBrowserProvider {
    private lateinit var controlsFragment: PlaybackControlsFragment
    override var mediaBrowser: MediaBrowser? = null
    private var mediaBrowserFuture: ListenableFuture<MediaBrowser>? = null

    private val playerListener = object : Player.Listener {
        override fun onPlaybackStateChanged(state: Int) {
            updateControlsVisibility()
        }

        override fun onMediaItemTransition(mediaItem: androidx.media3.common.MediaItem?, reason: Int) {
            updateControlsVisibility()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Timber.d("Activity onCreate")

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            @Suppress("DEPRECATION")
            val taskDesc = ActivityManager.TaskDescription(
                title.toString(),
                BitmapFactory.decodeResource(resources, R.drawable.ic_launcher_white),
                ContextCompat.getColor(this, R.color.primaryColor)
            )
            setTaskDescription(taskDesc)
        }
    }

    override fun onStart() {
        super.onStart()
        Timber.d("Activity onStart")

        controlsFragment = supportFragmentManager
            .findFragmentById(R.id.fragment_playback_controls) as? PlaybackControlsFragment
            ?: throw IllegalStateException(
                "Missing fragment with id 'controls'. Cannot continue."
            )

        hidePlaybackControls()
        connectBrowserIfNeeded()
    }

    override fun onStop() {
        super.onStop()
        Timber.d("Activity onStop")
        mediaBrowser?.removeListener(playerListener)
        mediaBrowser?.release()
        mediaBrowser = null
    }

    protected open fun onMediaControllerConnected() {
    }

    protected fun showPlaybackControls() {
        Timber.d("showPlaybackControls")
        if (NetworkHelper.isOnline(this)) {
            supportFragmentManager.beginTransaction()
                .setCustomAnimations(
                    R.animator.slide_in_from_bottom,
                    R.animator.slide_out_to_bottom,
                    R.animator.slide_in_from_bottom,
                    R.animator.slide_out_to_bottom
                )
                .show(controlsFragment)
                .commit()
        }
    }

    protected fun hidePlaybackControls() {
        Timber.d("hidePlaybackControls")
        supportFragmentManager.beginTransaction()
            .hide(controlsFragment)
            .commit()
    }

    private fun shouldShowControls(): Boolean {
        val browser = mediaBrowser ?: return false
        val hasItem = browser.currentMediaItem != null
        return hasItem &&
            browser.playbackState != Player.STATE_IDLE &&
            browser.playbackState != Player.STATE_ENDED
    }

    private fun updateControlsVisibility() {
        if (shouldShowControls()) {
            showPlaybackControls()
        } else {
            hidePlaybackControls()
        }
    }

    private fun connectBrowserIfNeeded() {
        if (mediaBrowser != null) {
            updateControlsVisibility()
            controlsFragment.onConnected()
            onMediaControllerConnected()
            return
        }
        val token = SessionToken(this, ComponentName(this, MusicService::class.java))
        val future = MediaBrowser.Builder(this, token).buildAsync()
        mediaBrowserFuture = future
        future.addListener(
            {
                try {
                    val browser = future.get()
                    mediaBrowser = browser
                    browser.addListener(playerListener)
                    updateControlsVisibility()
                    controlsFragment.onConnected()
                    onMediaControllerConnected()
                } catch (e: Exception) {
                    Timber.e(e, "Failed to connect MediaBrowser")
                    hidePlaybackControls()
                }
            },
            MoreExecutors.directExecutor()
        )
    }
}
